File: LeetCode/Explore/TopInterview/Blind75/DynamicProgramming/322_coin_change.py
class Solution:
    def coinChange(self, coins: List[int], amount: int) -> int:
        dp = [1000000] * (amount + 1)
        dp[0] = 0
        for i in range(1, amount + 1, 1):
          for coin in coins:
            if coin <= i and dp[i - coin] < 1000:
              dp[i] = min(dp[i], dp[i - coin] + 1)
              
        if dp[amount] <= amount:
          return dp[amount]
        else:
          return -1

File: LeetCode/Explore/TopInterview/Google/DynamicProgramming/322_coin_change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
      vector<int> dp(amount + 1, INT_MAX);
      dp[0] = 0;
      for (int money = 1; money <= amount; money++) {
        for (auto &coin : coins) {
          if (coin <= money && dp[money - coin] < INT_MAX) {
            dp[money] = min(dp[money], dp[money - coin] + 1);
          }
        }
      }
      return dp[amount] <= amount ? dp[amount] : -1;
    }
};
